src/demo-app/drag-and-drop/drag-and-drop-demo.ts

```diff
@@ -0,0 +1,499 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  Component,
+  Directive,
+  ContentChildren,
+  forwardRef,
+  QueryList,
+  ElementRef,
+  Inject,
+  Input,
+  Output,
+  EventEmitter,
+  ContentChild,
+  AfterContentInit,
+  ViewEncapsulation,
+} from '@angular/core';
+import {DOCUMENT} from '@angular/common';
+import {MatIconRegistry} from '@angular/material/icon';
+import {DomSanitizer} from '@angular/platform-browser';
+
+/** Event emitted when the user starts dragging an item in a drop container. */
+interface CdkDropStart<T> {
```

Comment:
I added the `transfer` and `remove` events so the consumer has a way of knowing when to remove the item from one array and add it to the other. We need these to fire once the user is done dragging, otherwise the consumer might swap out the data in an `ngFor` which will destroy the `CdkDrag` instance that is currently active. Are `CdkDragEnter` and `CdkDragExit` supposed to fire while the user is dragging or when they've dropped the item?

Also, technically the `remove` event isn't necessary since the `transfer` event has the data of both the source and destination drop containers. I put it in for consistency.

Reply:
* It sounds like `transfer` is just what I'm calling `drop`, except that it also happens when its within the same zone. Is there any reason you could do the remove on `drop`? As long as the user updates both data sources for the source and target locations in the same tick it should be okay.
* `enter` and `exit` would fire while you're still dragging

Reply:
What I meant is that the remove should only happen on `drop`, otherwise we risk it breaking while the user is still dragging, but I think that's what we agreed on previously.

Reply:
Yeah, agreed. Though we should try to gracefully fail when the user destroys the draggable in the middle of a drag.